### PR TITLE
Harden line directive filenames

### PR DESCRIPTION
**Pull request by shi-eric**
## Description

Escape source filenames before inserting them into generated C/CUDA `#line` directives. Previously Warp only normalized backslashes before interpolating the filename into a quoted preprocessor token, so quotes or control characters in a filename could break out of the directive and add generated source text.

This keeps the existing backslash-to-forward-slash normalization, preserves the existing `lineinfo` / debug-mode / `line_directives` emission gates, and adds focused coverage for quotes, LF, CR, tab, NUL, other control characters, DEL, and Windows-style paths.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://nvidia.github.io/warp/stable/user_guide/contribution_guide.html).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

## Test plan

- `WARP_CACHE_PATH=/tmp/warp-cache-line-directive-commit WARP_CACHE_ROOT=/tmp/warp-cache-line-directive-commit uv run warp/tests/test_codegen.py TestCodeGen.test_line_directive_escapes_filename TestCodeGen.test_line_directive_preserves_normalized_path`
- `WARP_CACHE_PATH=/tmp/warp-cache-line-directive-final WARP_CACHE_ROOT=/tmp/warp-cache-line-directive-final uv run warp/tests/test_codegen.py`
- `uvx pre-commit run --files warp/_src/codegen.py warp/tests/test_codegen.py`

## Bug fix

Without this PR, a filename containing a quote and newline can terminate the `#line` filename token and inject additional generated source lines:

```python
import linecache

import warp as wp
from warp._src.codegen import Adjoint

filename = '/tmp/warp_poc"\nint injected_from_filename;\n//.py'
source = "def kernel_fn(x: int):\n    y = x + 1\n    return y\n"
linecache.cache[filename] = (len(source), None, source.splitlines(True), filename)

namespace = {}
exec(compile(source, filename, "exec"), namespace)
adj = Adjoint(namespace["kernel_fn"])
adj.builder_options = {"lineinfo": True, "line_directives": True, "mode": "release"}

print(adj.get_line_directive("int x;", 0))
```

Before this PR, the printed directive contains physical newlines from the filename. With this PR, those characters are escaped inside the quoted filename token.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved robustness of code generation by enhancing filename handling in compiler directives to properly escape special characters (quotes, newlines, tabs, control characters) and normalize Windows-style paths, preventing potential compilation issues.

* **Tests**
  * Added tests for filename escaping and path normalization in code generation directives to ensure edge cases are properly handled.

### CODE REVIEW

File: warp/_src/codegen.py [hunk 1]
```diff
@@ -31,6 +31,28 @@
...
+            escaped.append('\\"')
+        elif c == "\n":
+            escaped.append("\\n")
+        elif c == "\r":
+            escaped.append("\\r")
+        elif c == "\t":
+            escaped.append("\\t")
+        elif ord(c) < 32 or ord(c) == 127:
+            # Use fixed-width octal so following filename characters cannot be consumed by the escape.
+            escaped.append(f"\\{ord(c):03o}")
```
[the remark this answers is not in the document]
Comment (shi-eric) [reply warp/_src/codegen.py:49]: @greptileai We are leaving non-ASCII Unicode filenames as-is here. The injection risk this PR fixes is filename content that can terminate or split the `#line` filename token: quotes and control characters. Valid non-ASCII Python filenames are Unicode strings and are written to generated source as UTF-8; octal-escaping U+0080 through U+00FF would change the debug filename bytes/meaning for normal Unicode paths rather than just hardening the directive syntax.

Raw non-UTF-8 filesystem bytes are represented in Python as surrogate code points, not ordinary 128-255 Unicode characters, so escaping only that range would not address that separate case. We can consider a distinct surrogate-path policy later if needed, but it is out of scope for this injection fix.

File: warp/tests/test_codegen.py [hunk 2]
```diff
@@ -1414,6 +1415,39 @@ def rebind_function_local_to_value_kernel(out: wp.array(dtype=float)):
...
+        adj = self._make_adjoint_with_filename("/tmp/warp_poc.py")
+        adj.filename = filename
+
+        directive = adj.get_line_directive("int x;", 0)
+
+        self.assertEqual(
+            directive,
+            '#line 1 "/tmp/warp_poc\\"\\n\\r\\t\\000\\037\\177int injected_from_filename;\\n//.py"',
+        )
+        self.assertEqual(directive.count("\n"), 0)
```
[the remark this answers is not in the document]
Comment (shi-eric) [reply warp/tests/test_codegen.py:1441]: @greptileai Not adding a high-byte test in this PR because we are intentionally not changing or pinning high-byte Unicode behavior here. The tests cover the directive injection surface this patch owns: quote, LF, CR, tab, NUL, other control characters, DEL, and Windows path normalization.

A test for U+0080 through U+00FF would either codify pass-through behavior that is not part of this fix, or imply octal escaping behavior that we do not want for valid Unicode filenames.
